x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/ConfigInitialNode.java

```diff
@@ -446,85 +458,95 @@ SecureString newKeystorePassword() {
     // Detect if the existing yml configuration is incompatible with auto-configuration,
     // in which case auto-configuration is SILENTLY skipped.
     // This assumes the user knows what she's doing when configuring the node.
-    void checkExistingConfiguration(Environment environment, Terminal terminal) throws UserException {
-        // Silently skipping security auto configuration, because Security is already configured.
-        if (environment.settings().hasValue(XPackSettings.SECURITY_ENABLED.getKey())) {
+    void checkExistingConfiguration(Settings settings, Terminal terminal) throws UserException {
+        // Silently skip security auto configuration when Security is already configured.
+        // Security is enabled implicitly, but if the admin chooses to enable it explicitly then
+        // skip the TLS auto-configuration, as this is a sign that the admin is opting for the default behavior
+        if (XPackSettings.SECURITY_ENABLED.exists(settings)) {
             // do not try to validate, correct or fill in any incomplete security configuration,
             // instead rely on the regular node startup to do this validation
-            terminal.println(expectedNoopVerbosityLevel(),
+            terminal.println(Terminal.Verbosity.VERBOSE,
                     "Skipping security auto configuration because it appears that security is already configured.");
             throw new UserException(ExitCodes.NOOP, null);
         }
-        // Silently skipping security auto configuration if enrollment is disabled.
-        // But tolerate enrollment explicitly enabled, as it could be useful to enable it by a command line option
-        // only the first time that the node is started.
-        if (environment.settings().hasValue(XPackSettings.ENROLLMENT_ENABLED.getKey()) && false ==
-                XPackSettings.ENROLLMENT_ENABLED.get(environment.settings())) {
-            terminal.println(expectedNoopVerbosityLevel(),
+        // Silently skip security auto configuration if enrollment is disabled.
+        if (XPackSettings.ENROLLMENT_ENABLED.exists(settings) && false == XPackSettings.ENROLLMENT_ENABLED.get(settings)) {
```

Comment:
I don't remember us discussing it.
I remember you mentioning that there ought be a setting to disable auto configuration.
I always assumed you were referring to the enrollment setting we have already.

As you describe, we can give the option to start with auto generated TLS but enrollment disabled. I think overall this adds little value because enrollment can be disabled at any later point in time.
Also, the condition that the node forms a single node cluster when it starts, is questionable if the goal is to only have TLS configured.

I'm open to discuss TLS only auto configuration, and implement it in a follow-u, but count me as skeptical. 

Overall I think this should remain a separate CLI to generate TLS for a node (and not a one invoked automatically)

Reply:
> I don't remember us discussing it.

I hope I'm not imagining things :). There was a discussion about a separate setting and I remember that we thought it was  a good idea and this is why I write this as is, not to presuppose a decision. We discuss many things daily though, it's understandable to not have the same recollection of all discussions.

> As you describe, we can give the option to start with auto generated TLS but enrollment disabled. 

This is not the end-goal. 

> I'm open to discuss TLS only auto configuration, and implement it in a follow-u, but count me as skeptical.

This is not my suggestion. 

I simply suggest to have a setting `xpack.security.auto_configuration.enabled` which users /tools /orchestration can expliitly set to false when they want to make sure that none of the auto-configuration runs. Yes, we do have the enrollment mode setting, but "As a user I don't find it clear that I need to disable enrollment mode to disable security auto configuration. Can't I just disable security auto-configuration?" 
We know it does the same thing, but users might not.

Reply:
I introduced the xpack.security.autoconfiguration.enabled setting and will be bailing if this is set to false, but I didn't touch the checking of the enrollment mode setting.  We can continue the discussion about security auto configuration without enrollment in a separate venue and adjust in a follow up

Reply:
We discussed and agreed that:

- `xpack.security.autoconfiguration.enabled` should control whether auto-configuration ( both TLS configuration and credential generation ) happens
- `xpack.security.enrollment_mode.enabled` should control whether Enrollment APIs (and enrollment related CLI tools ) are operational